### PR TITLE
Support "pending" option for Web3 requests

### PR DESCRIPTION
**Pull request by arajasek**
## Notice

It is not allowed to submit your PR to the master branch directly, please submit your PR to the master-pre-merge branch.

## Description

Adds the  "pending" option for the missing functions, thus fixing issue #709.

## Type of change

Insert **x** into the following checkboxes to confirm (eg. [x]):
- [ ] Bug fix.
- [ ] New feature.
- [x] Enhancement.
- [ ] Unit test.
- [ ] Breaking change (a fix or feature that causes existing functionality to not work as expected).
- [ ] Requires documentation update.

## Testing

All existing tests pass, new tests added.

## Verification

Insert **x** into the following checkboxes to confirm (eg. [x]):
- [x] I have self-reviewed my own code and conformed to the style guidelines of this project.
- [x] New and existing tests pass locally with my changes.
- [x] I have added tests for my fix or feature.
- [x] I have made appropriate changes to the corresponding documentation.
- [x] My code generates no new warnings.
- [x] Any dependent changes have been made.

### CODE REVIEW

File: modApiServer/src/org/aion/api/server/rpc/ApiWeb3Aion.java [hunk 1]
```diff
@@ -1131,13 +1095,13 @@ private FltrLg createFilter(ArgFltr rf) {
         Long bnFrom = parseBnOrId(rf.fromBlock);
         Long bnTo = parseBnOrId(rf.toBlock);
 
-        if (bnFrom == null || bnTo == null || bnFrom == -1 || bnTo == -1) {
+        if (bnFrom == null || bnTo == null) {
```
[the remark this answers is not in the document]
Comment (aion-kelvin) [reply modApiServer/src/org/aion/api/server/rpc/ApiWeb3Aion.java]: Looking at the code that comes after, I think that case is handled gracefully (just returns nothing)

File: modApiServer/src/org/aion/api/server/rpc/ApiWeb3Aion.java [hunk 2]
```diff
@@ -2872,6 +2838,14 @@ private IRepository getRepoByJsonBlockId(String _bnOrId) {
         return ac.getRepository().getSnapshotTo(b.getStateRoot());
     }
 
+    private AionBlock getBlockByBN(Long bn) {
```
Comment (aion-kelvin) [inline modApiServer/src/org/aion/api/server/rpc/ApiWeb3Aion.java]: I think long is better than Long here (probably doesn't make sense to pass in null)

File: modApiServer/src/org/aion/api/server/rpc/ApiWeb3Aion.java [hunk 3]
```diff
@@ -2872,6 +2838,14 @@ private IRepository getRepoByJsonBlockId(String _bnOrId) {
...
+        if (bn < 0) {
```
Comment (aion-kelvin) [inline modApiServer/src/org/aion/api/server/rpc/ApiWeb3Aion.java]: Should this not be `bn == -1` instead?

(And if so, I think it'd be good to put -1 in a const like 'BEST_PENDING_BLOCK' or something to that effect)

File: modApiServer/src/org/aion/api/server/rpc/ApiWeb3Aion.java [hunk 4]
```diff
@@ -2860,10 +2824,12 @@ public RpcMsg stratum_getMinerStats(Object _params) {
     private IRepository getRepoByJsonBlockId(String _bnOrId) {
         Long bn = parseBnOrId(_bnOrId);
         // if you passed in an invalid bnOrId, pending or it's an error
-        if (bn == null || bn < 0) {
+        if (bn == null) {
             return null;
         }
 
+        if (bn < 0) return pendingState.getRepository();
```
Comment (aion-kelvin) [inline modApiServer/src/org/aion/api/server/rpc/ApiWeb3Aion.java]: Should it not be bn == -1?

File: modApiServer/src/org/aion/api/server/rpc/ApiWeb3Aion.java [hunk 5]
```diff
@@ -535,17 +519,6 @@ public RpcMsg eth_getTransactionCount(Object _params) {
...
-        if (!bnOrId.equalsIgnoreCase("latest")) {
-            return new RpcMsg(
-                    null,
-                    RpcError.INVALID_PARAMS,
-                    "Default block parameter temporarily unsupported");
-        }*/
-
-        if (bnOrId.equalsIgnoreCase("pending")) {
-            return new RpcMsg(TypeConverter.toJsonHex(pendingState.bestPendingStateNonce(address)));
-        }
```
[the remark this answers is not in the document]
Comment (arajasek) [reply modApiServer/src/org/aion/api/server/rpc/ApiWeb3Aion.java]: Yes, they're equivalent.

File: modApiServer/src/org/aion/api/server/rpc/ApiWeb3Aion.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (arajasek) [reply modApiServer/src/org/aion/api/server/rpc/ApiWeb3Aion.java]: Yeah, it returns nothing, but I'll put in a special case and a log message so it's a little more robust.

File: modApiServer/src/org/aion/api/server/rpc/ApiWeb3Aion.java (hunk 4 above)
Comment (arajasek) [reply modApiServer/src/org/aion/api/server/rpc/ApiWeb3Aion.java]: Done

File: modApiServer/src/org/aion/api/server/rpc/ApiWeb3Aion.java (hunk 3 above)
Comment (arajasek) [reply modApiServer/src/org/aion/api/server/rpc/ApiWeb3Aion.java]: Done

File: modApiServer/src/org/aion/api/server/rpc/ApiWeb3Aion.java [hunk 6]
```diff
@@ -826,10 +788,12 @@ public RpcMsg eth_call(Object _params) {
         }
 
         Long bn = parseBnOrId(bnOrId);
-        if (bn == null || bn < 0) {
+        if (bn == null) {
```
[the remark this answers is not in the document]
Comment (arajasek) [reply modApiServer/src/org/aion/api/server/rpc/ApiWeb3Aion.java]: Good point! I added a clause to `parseBnOrId()` that returns `null` if it's given

**Comment by arajasek** [comment]
All suggested changes have been implemented :)

File: modApiServer/src/org/aion/api/server/rpc/ApiWeb3Aion.java [hunk 7]
```diff
@@ -1100,6 +1102,11 @@ private FltrLg createFilter(ArgFltr rf) {
             return null;
         }
 
+        if (bnTo != BEST_PENDING_BLOCK && (bnFrom > bnTo)) {
```
Comment (aion-kelvin) [inline modApiServer/src/org/aion/api/server/rpc/ApiWeb3Aion.java]: nit: unnecessary parens

Comment (arajasek) [reply modApiServer/src/org/aion/api/server/rpc/ApiWeb3Aion.java]: KILLED them :)

**Pull request merged**